hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -456,4 +517,113 @@ protected ICompletableFuture createCompletedFuture(Object value) {
                 clientContext.getExecutionService().getAsyncExecutor());
     }
 
+    protected final class NearCacheMembershipListener implements MembershipListener {
+
+        public void memberAdded(MembershipEvent event) {
+            MemberImpl member = (MemberImpl) event.getMember();
+            addInvalidationListener(member);
+        }
+
+        public void memberRemoved(MembershipEvent event) {
+            MemberImpl member = (MemberImpl) event.getMember();
+            removeInvalidationListener(member);
+        }
+
+        public void memberAttributeChanged(MemberAttributeEvent memberAttributeEvent) {
+
+        }
+
+    }
+
+    protected final class NearCacheInvalidationHandler implements EventHandler<CacheInvalidationMessage> {
+
+        private final Client client;
+
+        private NearCacheInvalidationHandler(Client client) {
+            this.client = client;
+        }
+
+        @Override
+        public void handle(CacheInvalidationMessage message) {
+            if (client.getUuid().equals(message.getSourceUuid())) {
+                return;
+            }
+            Data key = message.getKey();
+            if (key != null) {
+                nearCache.invalidate(key);
+            } else {
+                nearCache.clear();
+            }
+        }
+
+        @Override
+        public void beforeListenerRegister() {
+
+        }
+
+        @Override
+        public void onListenerRegister() {
+            nearCache.clear();
+        }
+
+    }
+
+    protected void registerInvalidationListener() {
+        if (nearCache != null && nearCache.isInvalidateOnChange()) {
+            ClientClusterService clusterService = clientContext.getClusterService();
+            nearCacheMembershipRegistrationId =
+                    clusterService.addMembershipListener(new NearCacheMembershipListener());
+            Collection<MemberImpl> memberList = clusterService.getMemberList();
+            for (MemberImpl member : memberList) {
+                addInvalidationListener(member);
+            }
+        }
+    }
+
+    protected void removeInvalidationListener() {
+        if (nearCache != null && nearCache.isInvalidateOnChange()) {
+            String registrationId = nearCacheMembershipRegistrationId;
+            ClientClusterService clusterService = clientContext.getClusterService();
+            if (registrationId != null) {
+                clusterService.removeMembershipListener(registrationId);
+            }
+            Collection<MemberImpl> memberList = clusterService.getMemberList();
+            for (MemberImpl member : memberList) {
+                removeInvalidationListener(member);
+            }
+        }
+    }
+
+    // TODO come up with a better sync
+    protected synchronized void addInvalidationListener(MemberImpl member) {
```

Comment:
@mdogan In this case, since `addInvalidationListener` is called my `MembershipListener` for new member, it should be `synchronized`. And also should be `synchronized` with `removeInvalidationListener`. Am I wrong ?